Feature: Unified Inline Diff Mechanism (No External Dependencies)

bassamsdata opened the pull request:
## Description

This PR introduces a new **inline diff mechanism** for codecompanion, implemented in the simplest way possible to maximize readability and maintainability. The diffing logic is based on the ZED style of diff, and is rendered in a unified inline style directly within the buffer. No external dependencies are required; the implementation uses only built-in Neovim APIs and Lua.

- **Unified Inline Diff**: Shows changes (additions and deletions including modified) inline, using extmarks and virtual text, similar to ZED's approach.
- **No Dependencies**: All logic is internal, avoiding any third-party libraries.
- **Simplicity**: The code is intentionally kept minimal and straightforward to ease future maintenance and reduce complexity.

### Notable Implementation Details (this WIP and could change a lot)

- **Diff Provider**: A new provider (`inline`) is added and set as the default. It can be toggled in the config.
- **Diff Logic**: The diff is calculated using Neovim's built-in `vim.diff` or `vim.text.diff` (if available with nvim 0.12), and visualized using extmarks for highlights and virtual lines for deletions.
- **Buffer Safety**: The diff is only shown if there are actual changes between the original and current buffer content.
- **Minimal State**: The diff object tracks only what is necessary (namespace, extmarks, etc.) for clarity and maintainability.

#### Notable Changes in `wait.lua`

- **Race Condition Fix**: The callback logic in `wait.lua` is now guarded by a `decision_made` flag and a `cleanup_and_callback` function. This ensures the callback is only executed once, preventing race conditions where both an autocmd and a timeout could trigger the callback.

---

> **Warning**
>
> This is a draft PR. The inline diff mechanism is working, but expect bugs and rough edges. The feature still needs significant work to be considered reliable and production-ready.

---

### TODO


- [ ] Make the inline diff mechanism more reliable and robust.
- [ ] Improve workflow predictability, especially reconciling differences between chat-based diffs (with tools) and the inline strategy.
- [ ] Refactor and improve highlight groups for better visual clarity.
- [ ] Implement a **"super diff buffer"** (inspired by ZED) that accumulates all changes, allowing users to view and revert all changes in a single clean diff buffer.
- [ ] Add a keymap to send all changes to the quickfix list (with line numbers).
- [ ] Tweak the visual appearance; optionally add a **"GitHub style"** diff if it can be done in under 10 lines of code.
- [ ] Add comprehensive tests for all diff scenarios and edge cases.
- [ ] Clean up the codebase, especially removing or reducing excessive logging.


## Related Issue(s)

[Here](https://github.com/olimorris/codecompanion.nvim/discussions/1607#discussioncomment-13436914) #1607

## Screenshots


<img width="1692" height="1764" alt="Screenshot 2025-07-21 at 11 02 05 PM" src="https://github.com/user-attachments/assets/6d673cc5-67a3-43a3-a39c-5746423c7a6e" />

### Comparing with Mini.diff

<details><summary>Screenshots</summary>
<p>



cc Inline mode:
<img width="2384" height="1364" alt="Screenshot_inline" src="https://github.com/user-attachments/assets/58175bf2-5dda-4358-a1b4-726b14fef05d" />

Mini.diff style (this can looks way messier but this is just a simple example:
<img width="2616" height="1322" alt="Screenshot_mini_diff" src="https://github.com/user-attachments/assets/88d13af4-7648-4c75-85e3-99977876d85c" />


</p>
</details> 

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [ ] I've updated `CodeCompanion.has` in the [init.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/init.lua#L239) file for my new feature
- [ ] I've added [test](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md#testing) coverage for this fix/feature
- [ ] I've updated the README and/or relevant docs pages
- [ ] I've run `make all` to ensure docs are generated, tests pass and my formatting is applied

olimorris commented:
This is like an early Christmas present.

olimorris commented:
This is looking awesome and I can't wait. Had a quick review of the code so far, great idea on the edit tracker, really, really nice and will scale brilliantly with some of the ideas I have for CodeCompanion in the near future.

I've been looking at the CodeCompanion internals even more, especially where files are kept, stored and named. Should the `super_diff.lua` go into its own folder outside of strategies? It feels like it's more global than being tied to a single strategy.

bassamsdata commented:
> This is looking awesome and I can't wait.

Thank you! I had hoped this PR would be ready by now, but I ran into some issues and wanted to make it more robust, especially with the edit_tracker. I was also out of town the last few days.

> Had a quick review of the code so far,

I hope the number of logs didn’t bother you too much! :)

> and will scale brilliantly with some of the ideas I have for CodeCompanion in the near future.

That’s great to hear! I think this will fit nicely with the "agent" mode you’re planning to implement.

> Should the super_diff.lua go into its own folder outside of strategies?

I agree. I’ll move it out of the strategies folder as well.

bassamsdata commented:
This PR is almost ready, but I expect to do a significant amount of work rebasing it. Just wanted to provide an update in case anyone stops by.

olimorris reviewed (dismissed):
<img width="2800" height="1200" alt="Image" src="https://github.com/user-attachments/assets/8f5c8b89-63ba-4466-8b3c-d1128ad14fab" />

Summary:
This PR introduces a new “Super Diff” feature to CodeCompanion.nvim, along with an enhanced inline diff provider, extensive logging, and full test coverage. It tracks all edits in a chat session, generates a consolidated markdown view of changes across multiple files/buffers, and allows accepting or rejecting all changes with keybindings. Many diff‐related utilities have been extracted or extended, and the default provider order has been updated to prefer the new inline provider.

Strengths:
- Super Diff implementation is comprehensive, with grouped edits, status indicators, code blocks, and highlights
- Inline and unified diff utilities are nicely modularized in providers/diff/&ast; and fully covered by tests
- End‐to‐end tests for edit tracking, diff creation, wait helper, and super diff buffer ensure reliability
- Logging calls provide detailed trace/debug output for easier troubleshooting
- Backward compatibility preserved: default diff provider order updated but users can still select mini_diff

Issues Found:
1. super_diff.lua (538 LOC): The single file is very large and mixes multiple responsibilities (markdown generation, highlights, keymaps, sticky header).  It would be easier to maintain if split into smaller modules/functions.
2. Performance concern: Extensive per‐hunk logging and deep table clones may impact responsiveness on large diffs. Consider conditional debug logging or batch operations.
3. Duplicate sentinel logic: The duplicate‐detection window uses hardcoded 1s threshold in several places. Factor into a shared constant or helper for consistency.
4. No top‐level module docstring or README update to explain the new super_diff feature and keybinding (`gD`).
5. Error paths in the new providers/diff/inline.lua and diff helpers rely on `pcall` but do not propagate errors; may swallow unexpected issues. Consider bubbling critical failures or notifying the user.
6. The `wait.for_decision` helper augroup name changed to `codecompanion.agent.tools.wait_`; ensure related documentation and migration notes are provided.

Suggestions:
- Refactor super_diff.lua by extracting markdown‐generation, highlight‐application, and UI setup into separate utility modules to improve readability and testability
- Add a new section in the README.md documenting the Super Diff feature, usage (`:CodeCompanionSuperDiff` or `gD`), and configuration options
- Introduce a shared constant (e.g. DIFF_DUPLICATE_WINDOW_NS) for the 1s duplicate‐detection threshold in edit_tracker
- Wrap verbose logging behind a debug‐only flag so it can be disabled in normal use
- Add error notifications for critical failures when loading providers or generating diffs
- Update CHANGELOG.md with a summary of breaking provider order change and new default inline provider

bassamsdata commented:
Thank you so much for the kind words and the tweaks, I’m really glad you liked the style!  

Regarding the super_diff, that’s on me. Since you refactored the tools and when I was _sleepy_ rebasing, I didn’t catch the changes or notice the new `orchestrator.lua` module. But now it’s working perfectly out of the box after any edit by pressing `gD`. I still want to improve the visuals a bit, so I’d really appreciate any feedback you have.  

For the naming and the current default, yes, I agree this should inline one should be the default. Maybe we could rename the current Neovim diff to `split_diff`.  
I’m not sure about removing it entirely, as that might upset some users. It’s hard to say how many prefer splits over inline diffs, especially now that we have the new builtin module with no plugins required.

I’ll take care of the other notes and plan to address them later tonight.  

> I noticed that rejected text moves with the cursor when in the inline diff:  

Nice catch! 😄 I think I know exactly why this is happening, it’s due to a hack I implemented with the `set_extmark` function for aesthetic reasons. I’ll try to fix it today, though the hack was because of limitation of Neovim’s function itself.  

One thing I’d love your advice or opinion on:  
I can’t hide my excitement about the PR you’re working on for ACP communication with CLI tools. Will it be possible to get the file paths that those tools edit, that is, to know which files they’re modifying?

If not, my assumption is that I’ll need to refactor the `edit_tracker` to use Git worktrees. This could significantly reduce the amount of code, but the downside is it would depend on Git. Of course, this would be for a separate PR, I just wanted to check since I had some trouble with the MCP `neovim_edit_file` external tool from MCPHUB figuring out which file it was editing automatically.

Thank you again!

olimorris commented:
> I can include this in this PR if you’d like!

I would love that.

BTW, could you drop me an email (my email is in my profile), would love to tell you what I have planned in the next couple of weeks and get your views.

olimorris commented:
The SuperDiff is blowing me away. Bravo 👏🏼. It looks incredible and is such a nice addition. If you're cool, I'll write a substantial section in the documents about it?

I earlier used the `insert_edit_into_file` tool on a file that wasn't open and received the following error:

```
...companion/strategies/chat/tools/catalog/helpers/diff.lua:122: Unknown option '/Users/Oli/Code/Neovim/codecompanion.nvim/quotes.txt'
```

I suspect this error caused the SuperDiff to display the changes as rejected:

<img width="3270" height="2134" alt="2025-08-11 12_56_56 - WezTerm@2x" src="https://github.com/user-attachments/assets/20cdd140-571d-4f0e-80cd-b6e18bbc3f33" />

A few other minor points:
- The `h2` header is obscured by render-markdown, which isn't the case in the chat buffer. Unsure what's causing that
- Would be great to be able to pass in local buffer opts to the superdiff like `wrap = true` from the config
- Do you think we should put the icons into the config as well and let users customize them?
- Do you think it's worth merging the `debug_window` config options with the SuperDiff ones? My thinking is that it reduces the configuration burden on the user and I think most would be happy if they share dimensions.

---

As you may well have seen in another PR that I'm working on, this diff feature will be incredibly important and the work you've done in `helpers/diff.lua` looks ideal. When you're happy with this, let me know and I'll merge asap.

olimorris commented on lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file.lua:
```diff
@@ -14,23 +15,20 @@ local api = vim.api
 ]] .. patch.prompt .. [[
 The system uses fuzzy matching and confidence scoring so focus on providing enough context to uniquely identify the location.
+
+REMEMBER: No *** Begin Patch / *** End Patch markers = FAILED EDIT! and this could led to you being fired.
```
Hahaha. Does this really work?

bassamsdata replied:
I was testing it yesterday 😄 but I’m not sure which part made GPT-4.1 work without any missing markers. That said, I’m not too keen on keeping it, so I’ll go ahead and remove it.

olimorris replied:
I've noticed GPT-4.1 has degraded, recently. It used to be super reliable with this algorithm

seblyng replied:
I can also confirm that I have a lot of problems with this using GPT-4.1 by default

bassamsdata replied:
I removed that phrase :) since I’ve been testing it since yesterday and haven’t missed any markers at all with the new prompt modification.

bassamsdata commented:
> If you're cool, I'll write a substantial section in the documents about it?

I’d love that! It would be fantastic to have a detailed section about SuperDiff in the documentation. Thank you so much!

Regarding the error, I believe it’s fixed now. As for the other feedback, I agree completely and have nearly finished addressing everything. I’m hoping to finalize the PR by tomorrow!

bassamsdata commented:
I’ve addressed all those issues, so everything should be fixed now (hopefuly). I did choose some names for the config options, but I’m not entirely sure if they’re the best. Notably, there’s a breaking change where `debug_window` was renamed to `debug_and_super_diff_window`, so I’d really appreciate your feedback on that.  

I also added a keymap `gt` for `always_accept`. The quickfix for accepted changes from the super diff should be working as well. Virtual text will now appear for diffs when using the keymap (with config option to remove them)

I wasn’t able to fix the header rendering with `render-markdown` in a clean way, even after reviewing the source code, so I changed how it’s displayed completely.  

At this point, I plan to add some small improvements and resolve any conflicts. Any feedback on potential errors would be greatly appreciated.  

Thank you!

olimorris commented:
> I’ve addressed all those issues, so everything should be fixed now (hopefuly). I did choose some names for the config options, but I’m not entirely sure if they’re the best. Notably, there’s a breaking change where debug_window was renamed to debug_and_super_diff_window, so I’d really appreciate your feedback on that.

I'll check if users have `debug_window` set and send them a deprecation notification advising them what to do. I'm trying to minimize breaking changes where possible. I'll have a think what else we could call the window options. Something like "info_window" or "child_window" with a comment explaining what it is.

> I also added a keymap gt for always_accept. The quickfix for accepted changes from the super diff should be working as well. Virtual text will now appear for diffs when using the keymap (with config option to remove them)

This is brilliant. Lovely idea.

> At this point, I plan to add a few small improvements and resolve any conflicts. Nonetheless, I expect to finalize it and have it ready to merge today (unless any errors arise). Any feedback on potential issues while testing would be greatly appreciated.

Fantastic. I'll give it a test later on tonight. Can't wait.

olimorris commented on lua/codecompanion/config.lua:
```diff
@@ -1032,11 +1048,18 @@ You must create or modify a workspace file through a series of prompts over mult
-      debug_window = {
-        ---@return number|fun(): number
+      -- Shared configuration for debug window and super diff window
+      debug_and_super_diff_window = {
```
I've decided that we should call this `child_window`. It implies that it's a window opened from a parent which will be the chat buffer

bassamsdata commented:
This is new to me, and that’s a good thing! I’ve been using it extensively for the past week without encountering this issue. I’m currently on the go, but once I’m back at my laptop, I’ll take a closer look and investigate.

olimorris commented:
It seems to be very hit and miss. The first time I tried to record it, it worked and I couldn't capture the error.

I wonder if _occasionally_ the keymaps are set too late. A `vim.schedule` call might remedy that. I'll take a look tomorrow and then I think this is ready for merging (once I write the docs).

bassamsdata commented:
> Enjoy your travel.

Thank you :)

I looked into your Neovim config and found the issue. You’re setting buffer-local keymaps for "ga" and "gr," which explains the unpredictable behavior, since there are two same buffer-local keymaps.

If these were global keymaps, there wouldn’t be any issues, and honestly, `vim.schedule` might help solve this problem. However, due to varying LSP speeds and asynchronous behavior, it can still be unpredictable (I think). I also noticed that after setting ours, buffer-local keymaps sometimes stop working following accept/reject actions. So, I pushed a fix that saves any existing buffer-local keymaps, deletes them while ours are active, and restores them when our keymaps are removed.

Other issues have been resolved as well. By the way, the `child_window` naming is very innovative and great, I really like it.

That said, I still have some improvements in mind, but I think we’re good to go for now once the docs are ready. I’ll follow up with a complementary PR right after this one to handle some nitpicks and add tests for the new features we’ve added recently.

Thanks again

olimorris commented:
Great spot regarding the keymaps. It wasn't an issue until now because the buffer had already been open and so the diff keymaps were set after my LSP ones. That gives me a couple of thoughts:

1. Are `ga` and `gr` bad defaults? Should they perhaps be `gda` and `gdr`?
2. Should we keep them as `ga` and `gr` but allow them to also work in the CodeCompanion buffer as a fallback. So the user can't do it in the edited buffer but can in CodeCompanion

I note that `ga` does seem like a common mapping for LSP code actions ☹️ so I'm thinking now could be a good chance to remap them.

I'll get cracking with the documentation today.

olimorris commented:
I've been thinking about the keymap situation. I don't expect users to be accepting or rejecting diffs very often within their workflow. As such, I think we've potentially overcomplicated the issue that I was experiencing. My preference would be we try and keep this PR as simple as possible, and if key maps clash, that would be an issue for the user to sort out. I propose we rollback 16c166f472ba0ef7edf94ca0049feef91af38da3 and set the default mappings to be `gda` and `gdr`.

I've made it clear in the documentation that users should be mindful of changing the defaults if they don't like it.

Otherwise, this is killer. I've played around a lot with this today and it's been incredible. Amazing job!

I've updated the docs and am working on an updated video for it.

bassamsdata commented:
I agree, my keymap implementation was a bit complicated, so leaving it up to the user is a good approach. Personally, I switched back immediately to `ga` and `gr` 😄.

I’ve reverted the commit and made some small enhancements with hardcoded values. Since this will require a force-push, I’ll wait until you’ve finished the changes you’re working on :)
